photutils/isophote/ellipse_model.pyx

```diff
@@ -0,0 +1,157 @@
+
+cnp.import_array()
+
+cdef extern from "math.h":
+
+    double cos(double x)
+    double sin(double x)
+    double sqrt(double x)
+
+
+DTYPE = np.float64
+ctypedef cnp.float64_t DTYPE_t
+
+cdef inline double get_intens_no_harmonics(double intens0, double phi, double a3, double b3, double a4, double b4):
+    return intens0
+
+cdef inline double get_intens_harmonics(double intens0, double phi, double a3, double b3, double a4, double b4):
+    return (
+        intens0
+        + a3 * sin(3.0 * phi)
+        + b3 * cos(3.0 * phi)
+        + a4 * sin(4.0 * phi)
+        + b4 * cos(4.0 * phi)
+    )
+
+
+def build_ellipse_model_c(
+    unsigned int n_rows,
+    unsigned int n_cols,
+    double phi_min,
+    double phi_max,
+    cnp.ndarray[DTYPE_t, ndim=1] finely_spaced_sma,
+    cnp.ndarray[DTYPE_t, ndim=1] intens_array,
+    cnp.ndarray[DTYPE_t, ndim=1] eps_array,
+    cnp.ndarray[DTYPE_t, ndim=1] pa_array,
+    cnp.ndarray[DTYPE_t, ndim=1] x0_array,
+    cnp.ndarray[DTYPE_t, ndim=1] y0_array,
+    cnp.ndarray[DTYPE_t, ndim=1] a3_array = None,
+    cnp.ndarray[DTYPE_t, ndim=1] b3_array = None,
+    cnp.ndarray[DTYPE_t, ndim=1] a4_array = None,
+    cnp.ndarray[DTYPE_t, ndim=1] b4_array = None,
+):
+    cdef cython.Py_ssize_t len_sma
+    len_sma = len(finely_spaced_sma)
+    for array in (intens_array, eps_array, pa_array, x0_array, y0_array):
+        if len(array) != len_sma:
+            raise ValueError(f"All input arrays must be same length={len_sma}")
+    harmonics_is_none = (
+        array is None for array in (a3_array, b3_array, a4_array, b4_array)
+    )
+
+    cdef double a3, b3, a4, b4
+
+    if all(harmonics_is_none):
+        intens_func = get_intens_no_harmonics
+        a3 = 0
+        b3 = 0
+        a4 = 0
+        b4 = 0
+        do_harmonics = False
+    else:
+        if any(harmonics_is_none):
+            raise ValueError("Must supply all harmonic arrays if any is not None")
+        for array in (a3_array, b3_array, a4_array, b4_array):
+            if len(array) != len_sma:
+                raise ValueError(f"All input arrays must be same length={len_sma}")
+        intens_func = get_intens_harmonics
+        do_harmonics = True
+
+    cdef double phi, fx, fy, x, y
+    cdef double one_m_fx, one_m_fy, one_m_fx_t_one_m_fy, one_m_fy_t_fx, one_m_fx_t_fy, fy_t_fx
+    cdef double r, sma, q, pa, x0, y0, intens0, intens
+    cdef int i, j, i_max, j_max
+    cdef cython.Py_ssize_t index
+
+    # Define output array
+    cdef double[:, :] result = np.zeros([n_rows, n_cols], dtype=DTYPE)
+    cdef double[:, :] weight = np.zeros([n_rows, n_cols], dtype=DTYPE)
+
+    i_max = n_cols - 1
+    j_max = n_rows - 1
+
+    for index in range(1, len_sma):
+        with cython.boundscheck(False):
+            sma = finely_spaced_sma[index]
+            q = 1.0 - eps_array[index]
+            pa = pa_array[index]
+            x0 = x0_array[index]
+            y0 = y0_array[index]
+            intens0 = intens_array[index]
+        phi = 0
```

Comment:
```suggestion
        phi = phi_min
```
I think this should be `phi_min` now?

Reply:
Actually I read it wrong, the existing code starts at phi=0 and then sets phi to `max((phi + 0.75 / r), phi_min)` for the next step. So in practice it's probably only actually a minimum for the first step, and I'm slightly more confused than before.